core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1865,6 +1876,39 @@ abstract class RDD[T: ClassTag](
   // RDD chain.
   @transient protected lazy val isBarrier_ : Boolean =
     dependencies.filter(!_.isInstanceOf[ShuffleDependency[_, _, _]]).exists(_.rdd.isBarrier())
+
+  /**
+   * Returns the random level of this RDD's computing function. Please refer to [[RDD.RandomLevel]]
+   * for the definition of random level.
+   *
+   * By default, an RDD without parents(root RDD) is IDEMPOTENT. For RDDs with parents, the random
+   * level of current RDD is the random level of the parent which is random most.
+   */
+  // TODO: make it public so users can set random level to their custom RDDs.
+  // TODO: this can be per-partition. e.g. UnionRDD can have different random level for different
+  // partitions.
+  private[spark] def computingRandomLevel: RDD.RandomLevel.Value = {
```

Comment:
We will need to expose this with `@Experimental` tag - cant keep it `private[spark]` given the implications for custom RDD's.